File: Rat In A Maze.py
#https://www.naukri.com/code360/problems/rat-in-a-maze_1215030

from os import *
from sys import *
from collections import *
from math import *

def searchMaze(arr, n):
    # Write your code here.
    
    #DFS
    def ways(arr,stack,result,i,j,visited):
        #base case
        if i<0 or j<0 or j>=len(arr[0]) or i>=len(arr):
            return (stack,result) 
        #if 0 can't use that path
        if arr[i][j]==0:
            return (stack,result)
        #destination
        if i==len(arr)-1 and j==len(arr[0])-1:
            if stack not in result:
                result.append(stack)
            return (stack,result)
        
        #if visited don't go through that path
        if visited[i][j]==1:
            return (stack,result) 

        #mark visited 
        visited[i][j]=1    
        
        down,result=ways(arr,stack+'D',result,i+1,j,visited)
        left,result=ways(arr,stack+'L',result,i,j-1,visited)
        right,result=ways(arr,stack+'R',result,i,j+1,visited)
        up,result=ways(arr,stack+'U',result,i-1,j,visited)

        
        #mark visited=0 to use it for some other path -> backtracking
        visited[i][j]=0
        return (stack,result)

    result=[]
    visited=[[0 for _ in range(len(arr[0]))]for _ in range(len(arr))]
    _,result=ways(arr,'',result,0,0,visited)
    return result

File: Rat in a Maze.cpp
//https://www.geeksforgeeks.org/problems/rat-in-a-maze-problem/1


class Solution {
  public:
    vector<string> result;
    vector<vector<int>> visited;
    void helper(int x,int y,vector<vector<int>>& maze, string tracker){
        if(x==(maze.size()-1) && y==(maze.size()-1)){
            result.push_back(tracker);
            return;
        }
        vector<pair<int,int>> directions={{1,0},{0,-1},{0,1},{-1,0}};
        int travel=0;
        for(auto &[nx,ny]: directions){
            int newX=x+nx, newY=y+ny;
            if(newX>-1 && newY>-1 && newX<maze.size() && newY<maze.size()){
                if(maze[newX][newY]==1 && visited[newX][newY]==-1){
                    visited[newX][newY]=1;
                    
                    if(travel==0) helper(newX, newY, maze, tracker+'D');
                    else if(travel==1) helper(newX, newY, maze, tracker+'L');
                    else if(travel==2) helper(newX, newY, maze, tracker+'R');
                    else if(travel==3) helper(newX, newY, maze, tracker+'U');
                    
                    visited[newX][newY]=-1;
                }
            }
            travel++;
        }
        
    }
    vector<string> ratInMaze(vector<vector<int>>& maze) {
        // code here
        visited.resize(maze.size(),vector<int> (maze.size(),-1));
        visited[0][0]=1;
        helper(0,0,maze,"");
        return result;
        
    }
};
